src/RecursiveIteration.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Traversable;
+
+/**
+ * Represents an iteration in a recursive iterator.
+ *
+ * @since [*next-version*]
+ */
+class RecursiveIteration extends AbstractBaseIteration implements RecursiveIterationInterface
+{
+    /**
+     * An array of path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $path;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|int|null                   $key   The iteration key.
+     * @param mixed                             $value The iteration value.
+     * @param string[]|Stringable[]|Traversable $path  A list of path segments.
+     */
+    public function __construct($key, $value, $path = [])
+    {
+        $this->_setKey($key)
+             ->_setValue($value)
+             ->_setPathSegments($path);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getPathSegments()
+    {
+        return $this->path;
+    }
+
+    /**
+     * Sets the path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[]|Traversable $path A list of path segments.
+     *
+     * @return $this
+     */
+    protected function _setPathSegments($path)
+    {
+        $this->path = $path;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getDepth()
+    {
+        return count($this->getPathSegments()) - 1;
```

Comment:
Disregard my previous comment. I got confused.

A path of 3 segments has a depth of 2. When calculating depth, the baseline is the root node of the tree. Given that the root node has a depth of zero, but the root can still have a key and a path segment, it follows that the depth is the number of path segments minus 1.

Once again, I will reference [the article](http://typeocaml.com/2014/11/26/height-depth-and-level-of-a-tree/) that outlines the concepts we agreed on.